pkg/rancher/rancher.go

```diff
@@ -618,3 +623,67 @@ func migrateEncryptionConfig(ctx context.Context, restConfig *rest.Config) error
 	}
 	return allErrors
 }
+
+// checks for deprecated RKE1 resources in the cluster to ensure that the cluster is not using any deprecated resources.
+func validateRKE1Resources(wranglerContext *wrangler.Context) error {
+	resources, err := checkForRKE1Resources(wranglerContext)
+	if err != nil {
+		return fmt.Errorf("checking for RKE1 resources: %w", err)
+	}
+	if len(resources) == 0 {
+		return nil
+	}
+
+	return fmt.Errorf("Rancher startup halted due to deprecated RKE1 resources detected:\n - %s\n\nRemove these resources to proceed", strings.Join(resources, "\n - "))
+}
+
+// checkForRKE1Resources scans for deprecated RKE1 (Rancher Kubernetes Engine v1) resources in the Rancher management context.
+func checkForRKE1Resources(wranglerContext *wrangler.Context) ([]string, error) {
+	var found []string
+
+	// Check for RKE1 clusters
+	clusters, err := wranglerContext.Mgmt.Cluster().List(metav1.ListOptions{})
+	if err != nil {
+		if k8serror.IsNotFound(err) {
+			// No clusters CRD installed, treat as no resources
+			clusters = &v3.ClusterList{}
+		} else {
+			return nil, fmt.Errorf("error checking RKE1 clusters: %w", err)
+		}
+	}
```

Comment:
```suggestion
	if k8serror.IsNotFound(err) {
		// No clusters CRD installed, treat as no resources
		clusters = &v3.ClusterList{}
	} else if err != nil {
			return nil, fmt.Errorf("error checking RKE1 clusters: %w", err)
	}
```

`k8serror.IsNotFound` performs a `nil` check, so this makes the logic a little simpler.

Though I have a clarifying question about the comment. Does the apiserver return a `NotFound` error if the CRD doesn't exist? If so I was not aware of that.